src/webpage/src/app.rs

```diff
@@ -0,0 +1,334 @@
+use std::collections::BTreeMap;
+use chrono::prelude::*;
+use eframe::egui::{CollapsingHeader, Context};
+use egui_plot::{Line, Plot, PlotPoints};
+use ewebsock::{connect, WsReceiver, WsSender};
+use humantime::format_duration;
+use url::Url;
+use web_sys::window;
+
+pub struct App {
+    receiver: WsReceiver,
+    sender: WsSender,
+    vehicles: BTreeMap<u8, BTreeMap<u8, BTreeMap<String, MessageInfo>>>,
+    search_query: String,
+    collapse_all: bool,
+    expand_all: bool,
+}
+
+#[derive(Clone)]
+struct FieldInfo {
+    history: Vec<(DateTime<Utc>, f64)>,
+    latest_value: f64,
+}
+
+#[derive(Clone)]
+struct MessageInfo {
+    last_sample_time: DateTime<chrono::Utc>,
+    fields: BTreeMap<String, FieldInfo>,
+}
+
+impl App {
+    pub fn new() -> Self {
+        let location = window().unwrap().location();
+        let host = location.host().unwrap();
+        let protocol = if location.protocol().unwrap() == "https:" {
+            "wss:"
+        } else {
+            "ws:"
+        };
+
+        let url = format!("{}//{}/mavlink/ws", protocol, host);
+
+        let (sender, receiver) = {
+            let url = Url::parse(&url)
+                .unwrap()
+                .to_string();
+            connect(url, ewebsock::Options::default()).expect("Can't connect")
+        };
+
+        Self {
+            receiver,
+            sender,
+            vehicles: Default::default(),
+            search_query: String::new(),
+            collapse_all: false,
+            expand_all: false,
+        }
+    }
+}
+
+impl eframe::App for App {
+    fn update(&mut self, ctx: &Context, _frame: &mut eframe::Frame) {
+        let now = Utc::now();
+        eframe::egui::CentralPanel::default().show(ctx, |ui| {
+            // Heading and search bar with buttons
+            ui.horizontal(|ui| {
+                ui.heading("Search:");
+                ui.add(
+                    eframe::egui::TextEdit::singleline(&mut self.search_query)
+                        .hint_text("Search..."),
+                );
+                if ui.button("Clear").clicked() {
+                    self.search_query.clear();
+                }
+                if ui.button("Collapse All").clicked() {
+                    self.collapse_all = true;
+                    self.expand_all = false;
+                }
+                if ui.button("Expand All").clicked() {
+                    self.expand_all = true;
+                    self.collapse_all = false;
+                }
+            });
+
+            // Process incoming messages
+            while let Some(message) = self.receiver.try_recv() {
+                if let ewebsock::WsEvent::Message(message) = message {
+                    if let ewebsock::WsMessage::Text(message) = message {
```

Comment:
```suggestion
                if let ewebsock::WsEvent::Message(ewebsock::WsMessage::Text(message)) = message {{
```

(then remember to remove this extra curl bracket)

Reply:
I also included the while